framework/src/org/checkerframework/framework/util/expressionannotations/ExpressionAnnotationHelper.java

```diff
@@ -0,0 +1,715 @@
+            if (type == null || p == null) {
+                return null;
+            }
+            Set<AnnotationMirror> replacement = AnnotationUtils.createAnnotationSet();
+            for (Class<? extends Annotation> vpa : expressionAnnos) {
+                AnnotationMirror anno = type.getAnnotation(vpa);
+                if (anno != null) {
+                    // Only replace annotations that might have been changed.
+                    replacement.add(anno);
+                }
+            }
+            p.replaceAnnotations(replacement);
+            if (type.getKind() != p.getKind()) {
+                // if the underlying types don't match, then this type has be substituted for a
+                // type variable, so don't recur. The primary annotation was copied because
+                // if the type variable might have had a primary annotation at a use.
+                // For example:
+                // <T> void method(@KeyFor("a") T t) {...}
+                // void use(@KeyFor("b") String s) {
+                //      method(s);  // the type of the parameter should be @KeyFor("a") String
+                // }
+                return null;
+            }
+            return super.scan(type, p);
+        }
+
+        @Override
+        protected Void compare(AnnotatedTypeMirror type, AnnotatedTypeMirror p) {
+            if (type == null || p == null) {
+                return null;
+            }
+            if (type.getKind() != p.getKind()) {
+                ErrorReporter.errorAbort("Should be the same. type: %s p: %s ", type, p);
+            }
+            return null;
+        }
+
+        @Override
+        protected Void combineRs(Void r1, Void r2) {
+            return null;
+        }
+    }
+
+    /**
+     * Whether or not atm has an expression annotation. If an annotated type does not have an
+     * expression annotation, then no standardization or viewpoint adaption is performed. (This
+     * check avoids calling time intensive methods unless absolutely required.)
+     */
+    private boolean hasExpressionAnnotation(AnnotatedTypeMirror atm) {
+        if (atm == null) return false;
+        Boolean b = new ExpressionAnnotationExists().visit(atm);
+        if (b == null) {
+            return false;
+        }
+        return b;
+    }
+
+    /** Checks whether or not an annotated type contains an expression annotation. */
+    private class ExpressionAnnotationExists extends AnnotatedTypeScanner<Boolean, Void> {
+        @Override
+        protected Boolean scan(AnnotatedTypeMirror type, Void aVoid) {
+            if (type == null) {
+                return false;
+            }
+            for (AnnotationMirror am : type.getAnnotations()) {
+                if (isExpressionAnno(am)) {
+                    return true;
+                }
+            }
+            return super.scan(type, aVoid);
+        }
+
+        boolean isExpressionAnno(AnnotationMirror am) {
```

Comment:
Can't you just re-use the already existing isExpressionAnno method? It looks like you have the same logic three times.